### ISSUE TITLE
Wii U core launching broken since 2022-08-05

### ISSUE THREAD

**Issue by vaguerant** [libretro/RetroArch#14300]
## Description

When using either **Load Core** or **Load Content** and selecting a file (with **Always Reload Core on Run Content ON**), RetroArch blackscreens on Wii U.

Core dump/strack trace from Wii U crash logs:

```
Core1: Instruction at 0x0101FF80 (from SRR0) invalid access of 0x6173209E (value from DAR)
00;01;30;136: 
--Proc15-Core1--------- OSContext 0x100CA9A0 --------------------

00;01;30;136: tag1  = 0x00000001 (expecting 0x4F53436F)
00;01;30;136: tag2  = 0x100B8C00 (expecting 0x6E747874)
00;01;30;136: TBR   = 0x00000001_45F1FD0B
00;01;30;136: CR    = 0x22000042
00;01;30;136: CTR   = 0x00000000
00;01;30;136: LR    = 0x0102058C coreinit.rpl|OSCreateAlarmEx+0xFFFFFA44
00;01;30;136: SRR0  = 0x0101FF80 coreinit.rpl|__OSZeroProcessMemory+0x1D4
00;01;30;136: SRR1  = 0x00005072

00;01;30;136: state = 0x000A

00;01;30;136: r0   = 0x01021118 (      16912664)  r16  = 0x00000000 (             0)
00;01;30;136: r1   = 0x100bd5d8 (     269211096)  r17  = 0x00000000 (             0)
00;01;30;136: r2   = 0x1050b000 (     273723392)  r18  = 0xa3f802a3 (   -1544027485)
00;01;30;136: r3   = 0x61732062 (    1634934882)  r19  = 0x10080000 (     268959744)
00;01;30;136: r4   = 0x61732062 (    1634934882)  r20  = 0x10040000 (     268697600)
00;01;30;136: r5   = 0x61732062 (    1634934882)  r21  = 0x10040000 (     268697600)
00;01;30;136: r6   = 0x72750000 (    1920270336)  r22  = 0x10078a60 (     268929632)
00;01;30;136: r7   = 0x00000001 (             1)  r23  = 0x10080000 (     268959744)
00;01;30;136: r8   = 0x100cece0 (     269282528)  r24  = 0x00000000 (             0)
00;01;30;136: r9   = 0x00000001 (             1)  r25  = 0x00000000 (             0)
00;01;30;136: r10  = 0x01031168 (      16978280)  r26  = 0x00000000 (             0)
00;01;30;136: r11  = 0x61732062 (    1634934882)  r27  = 0x100ce824 (     269281316)
00;01;30;136: r12  = 0x10055e80 (     268787328)  r28  = 0x00000000 (             0)
00;01;30;136: r13  = 0x1050b000 (     273723392)  r29  = 0x10010000 (     268500992)
00;01;30;136: r14  = 0x00000000 (             0)  r30  = 0x61732062 (    1634934882)
00;01;30;136: r15  = 0x00000000 (             0)  r31  = 0x1099c868 (     278513768)
00;01;30;136: 
--Stack Trace--------------------------
00;01;30;137: 
Address:      Back Chain    LR Save
00;01;30;137: 0x100bd5d8:   0x100bd5e0    0x100ce4e8 coreinit.rpl|environ+0x6f27c
00;01;30;137: 0x100bd5e0:   0x100bd5f8    0x01021118 coreinit.rpl|OSCancelAlarm+0xdc
00;01;30;137: 0x100bd5f8:   0x100bd610    0x010311e4 coreinit.rpl|OSSignalEvent+0xac
00;01;30;137: 0x100bd610:   0x100bd648    0x010518ac coreinit.rpl|IPCDriverDisplayStatistics+0xf7c
00;01;30;137: 0x100bd648:   0x100bd670    0x01052e78 coreinit.rpl|IPCDriverDisplayStatistics+0x2548
00;01;30;137: 0x100bd670:   0x00000000    0xfff02b08 <unknown>+0x0
00:01:30:266: COS encountered unrecoverable error...

Pending Resource Request for COS processes...
        Title Id       Process Id -> RM Path             
```

### Expected behavior

Core should load as normal (if using **Load Core**) and start running the loaded content (if using **Load Content**).

### Actual behavior

The console locks up with a black screen, forcing a hard shut down via power button.

### Steps to reproduce the bug

1. Launch RetroArch
2. Load Core

**or**

1. Launch RetroArch
2. Make sure **Always Reload Core on Run Content** is enabled under **Settings > Core**
3. Load Content

### Bisect Results
Core launching was working normally in [2022-08-04 nightly](https://buildbot.libretro.com/nightly/nintendo/wiiu/2022-08-04_RetroArch.7z) but broken in [2022-08-05 nightly](https://buildbot.libretro.com/nightly/nintendo/wiiu/2022-08-05_RetroArch.7z)

Those nightlies span these 20 commits:
https://github.com/libretro/RetroArch/compare/88bc26d...ef55147

### Version/Commit
You can find this information under Information/System Information

- RetroArch: ef55147 or later

### Environment information

- OS: Wii Menu 5.5.5 (Europe), coldboot Tiramisu homebrew environment

**Comment by LibretroAdmin** [comment]
Bisected it further

**Comment by vaguerant** [comment]
The latest nightly is currently 3dc9793, one commit behind as of this comment, and does exhibit the issue. I don't have a RetroArch Wii U build environment set up currently to test the very latest.

**Comment by LibretroAdmin** [comment]
You really need to set it up then, I don't have one setup either.

**Comment by LibretroAdmin** [comment]
I can tell you the commits which it most likely won't be -
* 6ad88b080ec9f8bf776109f5f06d3d9242c4481e - highly unlikely
* 993aaee716502a30e7d0c2453591c214c79917b9 - unlikely
* 7068eebe0c59fea26a4fc176c769ef46fd3ad718 - unlikely, not WiiU-related
* 67c23aec0e7abfceb2f394857cf014d40693df28 - unlikely, just adds a return value
* 2e5df6a9738de42d1c57e86143b768d34c851447 - unlikely
* e1caf32817df57895f43b8667e7072790659e48d - unlikely, not WiiU-related
* 0594df8cce765acfc3e911b7d83f18eff1ee6d6e - unlikely
* 72a7b3ce0714ec756de3a77384cc0b7fc09760d0 - unlikely, not WiiU-related
* 9c7b96b9d0514015a4ecbb1bb45bb81bfed40d85 - unlikely
* e3b82184a705ac8e4a9b426210b89199c570dc3a - unlikely, doesn't change anything for WiiU
* b04cb4b1e50ff1209d92c61da9976a5e4156eade - unlikely, not WiiU-related

**Comment by vaguerant** [comment]
All right, it looks to be 3aa1811, the very first commit of that range, which is causing issues on Wii U.

**Comment by LibretroAdmin** [comment]
Oh, that's interesting. How would that affect the core starting up?

@Cthulhu-throwaway Any ideas?

**Comment by vaguerant** [comment]
> You're confident it's that commit and the commit prior to that is fine?

Yep, the commit immediately before (88bc26d) does not exhibit the issue. I also pulled the current HEAD and reverted 3aa1811 and its direct descendants:
- c6b99a5
- add4a1e
- b04cb4b
- a7c93eb

That build also doesn't exhibit the issue.

**Comment by LibretroAdmin** [comment]
OK, @Cthulhu-throwaway will likely give you some patches to test to see if we can get WiiU up and running again.

**Comment by Cthulhu-throwaway** [comment]
Strange, I didn't do anything to Wii U other than add some macros from libcafe (mainly to properly support getnameinfo).

Does it happen on a clean launch without doing anything network related (netplay, fetch core list, update cores, etc)?

**Comment by Cthulhu-throwaway** [comment]
Comparing WiiU related from that PR and before it:

https://github.com/libretro/RetroArch/blob/3aa1811ca911d39ef8eac3a37dff12291a119a99/libretro-common/net/net_compat.c#L152-L172 vs https://github.com/libretro/RetroArch/blob/88bc26da4e947aefa4099f757bddfd9770768113/libretro-common/net/net_compat.c#L238-L259

https://github.com/libretro/RetroArch/blob/3aa1811ca911d39ef8eac3a37dff12291a119a99/libretro-common/net/net_compat.c#L186-L196 vs https://github.com/libretro/RetroArch/blob/88bc26da4e947aefa4099f757bddfd9770768113/libretro-common/net/net_compat.c#L179-L189

https://github.com/libretro/RetroArch/blob/3aa1811ca911d39ef8eac3a37dff12291a119a99/libretro-common/net/net_compat.c#L459-L488 vs https://github.com/libretro/RetroArch/blob/88bc26da4e947aefa4099f757bddfd9770768113/libretro-common/net/net_compat.c#L332-L352

I don't see any difference in the behavior of the code.

**Comment by Cthulhu-throwaway** [comment]
net_compat_thread probably needs to be kept in the heap, if that address is accessed later on (which appears to be the case).

**Comment by LibretroAdmin** [comment]
@vaguerant care to try the latest commit?

**Comment by vaguerant** [comment]
Can confirm that this is resolved by 078f712. Tested on 5ffe62a and the issue no longer occurs. Thanks, @Cthulhu-throwaway!
